### PR TITLE
[HINTS] Set add hint

### PR DESCRIPTION
**Pull request by Juan-M-V**
# Set add hint implementation

## Description
Implements hint set_add: https://github.com/starkware-libs/cairo-lang/blob/2abd303e1808612b724bc1412b2b5babd04bb4e7/src/starkware/cairo/common/set.cairo#L34-L45

## Checklist
- [x] Unit tests added
- [x] Integration tests added.

### CODE REVIEW

File: src/vm/vm_memory/memory.rs [hunk 1]
```diff
@@ -113,6 +113,20 @@ impl Memory {
...
+        size: usize,
+    ) -> Result<Vec<Option<&MaybeRelocatable>>, MemoryError> {
+        let mut values = Vec::new();
+
+        for i in 0..size {
+            values.push(self.get(&addr.add_usize_mod(i, None))?);
+        }
+
+        Ok(values)
+    }
```
Comment (Oppen) [inline src/vm/vm_memory/memory.rs:129]: I'd much rather have an iterator return elements one by one. But even if we keep it this way, you should probably do it like this:
```suggestion
    pub fn get_range(
        &self,
        addr: &MaybeRelocatable,
        size: usize,
    ) -> Result<Vec<Option<&MaybeRelocatable>>, MemoryError> {
        let addr = if let RelocatableValue(addr) = addr {
            addr
        } else {
            return // some error; we'll write a helper eventually
        }
        let segment = self.data[addr.segment_index];
        let (i, n) = (addr.offset, addr.offset + size);
        if n > segment.len() {
            return // some error
        }
        Ok(Vec::from(segment[i..n]));
    }
```
It's fine to access `data` from within the `memory` module. The problem is only if we do it from the outside.

File: src/vm/hints/set.rs [hunk 2]
```diff
@@ -139,80 +39,72 @@ pub fn set_add(
...
+                .as_any()
+                .downcast_ref::<RangeCheckBuiltinRunner>()
+                .ok_or(VirtualMachineError::NoRangeCheckBuiltin)?;
+            let elm_size = if let MaybeRelocatable::Int(ref elm_size) = maybe_rel_elm_size {
+                if elm_size.is_zero() {
+                    return Err(VirtualMachineError::ValueNotPositive(elm_size.clone()));
+                }
+                elm_size
+                    .to_usize()
+                    .ok_or_else(|| VirtualMachineError::ValueNotPositive(elm_size.clone()))?
```
Comment (Oppen) [inline src/vm/hints/set.rs:57]: This could also mean you passed a negative number.

File: src/vm/hints/set.rs [hunk 3]
```diff
@@ -139,80 +39,72 @@ pub fn set_add(
...
+                    set_end_ptr.clone(),
+                ));
             }
-            Err(VirtualMachineError::NoRangeCheckBuiltin)
+
+            let set_span = set_end_ptr.sub(set_ptr, &vm.prime)?;
+            // sub method always returns a MaybeRelocatable::Int
+            let range_limit = if let MaybeRelocatable::Int(ref range_limit) = set_span {
+                range_limit
+                    .to_i32()
```
Comment (Oppen) [inline src/vm/hints/set.rs:80]: Shouldn't this be `to_usize`? Either that or you're returning the wrong error.

File: src/vm/hints/set.rs (hunk 2 above)
Comment (Juan-M-V) [reply src/vm/hints/set.rs:57]: I can't find the code but to_usize [returns None](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3ca52cc031a19f887154c56a6f2c57ab) if the given BigInt is negative

File: src/vm/hints/set.rs (hunk 3 above)
Comment (Juan-M-V) [reply src/vm/hints/set.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/289/commits/6463e1761b015ebf6c0c867d02673a7aaf105dda

File: src/vm/vm_memory/memory.rs (hunk 1 above)
Comment (Juan-M-V) [reply src/vm/vm_memory/memory.rs:129]: I can't resolve this trivially. Do you think I can change this in a different PR and make a proper iterator?

Comment (Oppen) [reply src/vm/vm_memory/memory.rs:129]: OK.

File: src/vm/hints/set.rs (hunk 2 above)
Comment (Oppen) [reply src/vm/hints/set.rs:57]: Exactly. But it also returns `None` if it's greater than `usize::MAX`, [doesn't it](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3ca52cc031a19f887154c56a6f2c57ab)?
The error message is not considering that case.

Comment (Juan-M-V) [reply src/vm/hints/set.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/289/commits/4f77a8115240fada313453bcdccd79776a81675b

Comment (Oppen) [reply src/vm/hints/set.rs]: I meant "too big" :facepalm:

File: src/vm/hints/set.rs [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+            )
+            .expect("Unexpected memory insert fail");
+        vm.memory
+            .insert(
+                &MaybeRelocatable::from((2, 0)),
+                if let Some(rel) = elm_a {
+                    rel
+                } else {
+                    &elm_a_default
+                },
```
Comment (Oppen) [inline src/vm/hints/set.rs:215]: You can use https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or

**Comment by Oppen** [comment]
LGTM. Just some minor refactor by using [unwrap_or](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or) for tests and it's good to go.

File: src/vm/hints/set.rs (hunk 4 above)
Comment (Juan-M-V) [reply src/vm/hints/set.rs]: resolve: https://github.com/lambdaclass/cleopatra_cairo/pull/289/commits/0f008d1c33335380eb9dc3b7f52a9aededd12905

**Review by Oppen — APPROVED** [review]

**Pull request merged**